sungrow/sungrowdiscovery.cpp

```diff
@@ -0,0 +1,161 @@
+*
+* GNU Lesser General Public License Usage
+* Alternatively, this project may be redistributed and/or modified under the
+* terms of the GNU Lesser General Public License as published by the Free
+* Software Foundation; version 3. This project is distributed in the hope that
+* it will be useful, but WITHOUT ANY WARRANTY; without even the implied
+* warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this project. If not, see <https://www.gnu.org/licenses/>.
+*
+* For any further details and any questions please contact us under
+* contact@nymea.io or see our FAQ/Licensing Information on
+* https://nymea.io/license/faq
+*
+* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include "sungrowdiscovery.h"
+#include "extern-plugininfo.h"
+
+SungrowDiscovery::SungrowDiscovery(NetworkDeviceDiscovery *networkDeviceDiscovery, quint16 port, quint16 modbusAddress, QObject *parent) :
+    QObject{parent},
+    m_networkDeviceDiscovery{networkDeviceDiscovery},
+    m_port{port},
+    m_modbusAddress{modbusAddress}
+{
+
+}
+
+void SungrowDiscovery::startDiscovery()
+{
+    qCInfo(dcSungrow()) << "Discovery: Start searching for Sungrow inverters in the network";
+    m_startDateTime = QDateTime::currentDateTime();
+
+    NetworkDeviceDiscoveryReply *discoveryReply = m_networkDeviceDiscovery->discover();
+    connect(discoveryReply, &NetworkDeviceDiscoveryReply::networkDeviceInfoAdded, this, &SungrowDiscovery::checkNetworkDevice);
+    connect(discoveryReply, &NetworkDeviceDiscoveryReply::finished, discoveryReply, &NetworkDeviceDiscoveryReply::deleteLater);
+    connect(discoveryReply, &NetworkDeviceDiscoveryReply::finished, this, [=] () {
+        qCDebug(dcSungrow()) << "Discovery: Network discovery finished. Found" << discoveryReply->networkDeviceInfos().count() << "network devices";
+
+        // Give the last connections added right before the network discovery finished a chance to check the device...
+        QTimer::singleShot(3000, this, [this] () {
+            qCDebug(dcSungrow()) << "Discovery: Grace period timer triggered.";
+            finishDiscovery();
+        });
+    });
+}
+
+QList<SungrowDiscovery::SungrowDiscoveryResult> SungrowDiscovery::discoveryResults() const
+{
+    return m_discoveryResults;
+}
+
+void SungrowDiscovery::checkNetworkDevice(const NetworkDeviceInfo &networkDeviceInfo)
+{
+    /* Create a Sungrow connection and try to initialize it.
+    * Only if initialized successfully and all information have been fetched correctly from
+    * the device we can assume this is what we are locking for (ip, port, modbus address, correct registers).
+    */
+
+    qCInfo(dcSungrow()) << "Creating Sungrow Modbus TCP connection for" << networkDeviceInfo.address() << "Port:" << m_port << "Slave Address" << m_modbusAddress;
```

Comment:
Please use `qCDebug` in this class. The only things which might be interesting in `qCInfo` might be: discovery started, discovery stopped, found n devices. In the `IntegrationPluginSungrow::discoverThings` you already inform about the discovered inverters.

In addition it is very helpful for debugging if you use within this class the Syntax:
`qCDebug(dcSungrow()) << "Discovery: ..."`

This allows to grep the logs only for the discovery process and makes it easier to follow the output if there are issues `journalctl -f -u nymead | grep "Sungrow: Discovery:"`.